### ISSUE TITLE
Regression: Generic decorator breaks the inference of complex types with unions

### ISSUE THREAD

**Issue by minmax** [microsoft/pyright#7722]
After update to pyright 1.1.359 there were errors in the previously working code.

```python
from __future__ import annotations

from typing import TYPE_CHECKING
from typing import Any
from typing import Generic
from typing import ParamSpec
from typing import TypeAlias
from collections.abc import Callable
from typing_extensions import TypeVar

P = ParamSpec("P")
T = TypeVar("T", infer_variance=True)
E = TypeVar("E", infer_variance=True)
R = TypeVar("R", infer_variance=True)
RE = TypeVar("RE", infer_variance=True)

class Ok(Generic[T]):
    pass

class Err(Generic[E]):
    pass

Result: TypeAlias = Ok[T] | Err[E]


def bind(func: Callable[[T], Result[R, RE]]) -> Callable[[Result[T, E]], Result[R, RE | E]]:
    def bind_wrapper(res: Result[T, E]) -> Result[R, RE | E]: ...

    # 1.1.359 error: Function return type "Result[R@bind, RE@bind | E@bind_wrapper]" is incompatible with type "Result[R@bind, RE@bind | E@bind]
    return wrap_function(bind_wrapper, func)


def wrap_function(wrapper: Callable[P, T], wrapped: Callable[..., Any]) -> Callable[P, T]: ...
```

and pyright is ok, when `wrap_function` was removed

```python
def bind(func: Callable[[T], Result[R, RE]]) -> Callable[[Result[T, E]], Result[R, RE | E]]:
    def bind_wrapper(res: Result[T, E]) -> Result[R, RE | E]: ...

    return bind_wrapper  # 1.1.359 ok
```

**Comment by erictraut** [comment]
This issue originates from a change I made in response to [this issue](https://github.com/microsoft/pyright/issues/4285). In retrospect, that was a bad change, and I've decided to reverse it. It wasn't really compliant with the typing spec, and it caused side effects that I've been working around ever since. Backing out this change solves a bunch of issues (including this one), removes a bunch of complexity, and makes pyright more compliant with the typing spec.
